File: android/app/src/main/java/com/claudelists/app/api/RealtimeClient.kt
package com.claudelists.app.api

import android.util.Log
import com.claudelists.app.BuildConfig
import io.ktor.client.*
import io.ktor.client.engine.okhttp.*
import io.ktor.client.plugins.websocket.*
import io.ktor.websocket.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*

private const val TAG = "RealtimeClient"

/**
 * WebSocket client for realtime database change notifications.
 *
 * Subscribes to postgres_changes events for:
 * - case_status: Status changes
 * - comments: Comment additions/deletions
 * - notifications: New notifications for watched cases
 * - watched_cases: Watch list changes
 *
 * Event types defined in supabase/functions/_shared/types.ts
 */
class RealtimeClient(private val authManager: AuthManager) {

    private val json = Json { ignoreUnknownKeys = true }

    private val client = HttpClient(OkHttp) {
        install(WebSockets)
    }

    private val baseUrl = BuildConfig.API_BASE_URL
        .replace("https://", "wss://")
        .replace("http://", "ws://")
    private val anonKey = BuildConfig.API_ANON_KEY

    private var session: WebSocketSession? = null
    private var connectionJob: Job? = null
    private var heartbeatJob: Job? = null
    private var messageRef = 0

    // Event flows
    private val _statusChangeEvents = MutableSharedFlow<PostgresChange>()
    val statusChangeEvents: SharedFlow<PostgresChange> = _statusChangeEvents

    private val _commentChangeEvents = MutableSharedFlow<PostgresChange>()
    val commentChangeEvents: SharedFlow<PostgresChange> = _commentChangeEvents

    private val _notificationEvents = MutableSharedFlow<NotificationChange>()
    val notificationEvents: SharedFlow<NotificationChange> = _notificationEvents

    private val _watchedCaseEvents = MutableSharedFlow<PostgresChange>()
    val watchedCaseEvents: SharedFlow<PostgresChange> = _watchedCaseEvents

    private val _connectionState = MutableStateFlow(ConnectionState.DISCONNECTED)
    val connectionState: StateFlow<ConnectionState> = _connectionState

    enum class ConnectionState {
        DISCONNECTED, CONNECTING, CONNECTED, ERROR
    }

    /**
     * Connect to realtime server and subscribe to changes.
     */
    suspend fun connect() {
        if (_connectionState.value == ConnectionState.CONNECTED) {
            Log.d(TAG, "Already connected")
            return
        }

        _connectionState.value = ConnectionState.CONNECTING

        val accessToken = authManager.getAccessToken()
        if (accessToken == null) {
            Log.e(TAG, "No access token, cannot connect")
            _connectionState.value = ConnectionState.ERROR
            return
        }

        val wsUrl = "$baseUrl/realtime/v1/websocket?apikey=$anonKey&vsn=1.0.0"

        try {
            connectionJob = CoroutineScope(Dispatchers.IO).launch {
                client.webSocket(wsUrl) {
                    session = this
                    Log.i(TAG, "WebSocket connected")
                    _connectionState.value = ConnectionState.CONNECTED

                    // Send access token
                    sendMessage(
                        topic = "realtime:system",
                        event = "access_token",
                        payload = buildJsonObject {
                            put("access_token", accessToken)
                        }
                    )

                    // Join the postgres_changes channel
                    joinChannel(accessToken)

                    // Start heartbeat
                    heartbeatJob = launch {
                        while (isActive) {
                            delay(30_000)
                            sendHeartbeat()
                        }
                    }

                    // Process incoming messages
                    for (frame in incoming) {
                        when (frame) {
                            is Frame.Text -> handleMessage(frame.readText())
                            is Frame.Close -> {
                                Log.i(TAG, "WebSocket closed")
                                _connectionState.value = ConnectionState.DISCONNECTED
                            }
                            else -> {}
                        }
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "WebSocket connection failed", e)
            _connectionState.value = ConnectionState.ERROR
        }
    }

    private suspend fun WebSocketSession.joinChannel(accessToken: String) {
        // Subscribe to all tables we care about
        val tables = listOf("case_status", "comments", "notifications", "watched_cases")

        sendMessage(
            topic = "realtime:postgres_changes",
            event = "phx_join",
            payload = buildJsonObject {
                put("config", buildJsonObject {
                    put("postgres_changes", buildJsonArray {
                        tables.forEach { table ->
                            add(buildJsonObject {
                                put("event", "*")
                                put("schema", "public")
                                put("table", table)
                            })
                        }
                    })
                })
                put("access_token", accessToken)
            }
        )

        Log.i(TAG, "Joined postgres_changes channel for tables: $tables")
    }

    private suspend fun WebSocketSession.sendMessage(
        topic: String,
        event: String,
        payload: JsonObject
    ) {
        val message = buildJsonObject {
            put("topic", topic)
            put("event", event)
            put("payload", payload)
            put("ref", (++messageRef).toString())
        }
        send(Frame.Text(message.toString()))
    }

    private suspend fun sendHeartbeat() {
        session?.sendMessage(
            topic = "phoenix",
            event = "heartbeat",
            payload = buildJsonObject {}
        )
    }

    private suspend fun handleMessage(text: String) {
        try {
            val message = json.parseToJsonElement(text).jsonObject
            val topic = message["topic"]?.jsonPrimitive?.content ?: return
            val event = message["event"]?.jsonPrimitive?.content ?: return
            val payload = message["payload"]?.jsonObject ?: return

            Log.d(TAG, "Received: topic=$topic, event=$event")

            when (event) {
                "postgres_changes" -> handlePostgresChange(payload)
                "phx_reply" -> handleReply(payload)
                "system" -> Log.d(TAG, "System message: $payload")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse message: $text", e)
        }
    }

    private suspend fun handlePostgresChange(payload: JsonObject) {
        val data = payload["data"]?.jsonObject ?: return
        val table = data["table"]?.jsonPrimitive?.content ?: return
        val type = data["type"]?.jsonPrimitive?.content ?: return
        val record = data["record"]?.jsonObject
        val oldRecord = data["old_record"]?.jsonObject

        Log.d(TAG, "Postgres change: $table $type")

        val change = PostgresChange(
            table = table,
            type = type,
            record = record,
            oldRecord = oldRecord
        )

        when (table) {
            "case_status" -> _statusChangeEvents.emit(change)
            "comments" -> _commentChangeEvents.emit(change)
            "notifications" -> {
                if (type == "INSERT" && record != null) {
                    val notification = parseNotification(record)
                    if (notification != null) {
                        _notificationEvents.emit(notification)
                    }
                }
            }
            "watched_cases" -> _watchedCaseEvents.emit(change)
        }
    }

    private fun parseNotification(record: JsonObject): NotificationChange? {
        return try {
            NotificationChange(
                type = record["type"]?.jsonPrimitive?.content ?: "",
                caseNumber = record["case_number"]?.jsonPrimitive?.content ?: "",
                listSourceUrl = record["list_source_url"]?.jsonPrimitive?.content ?: "",
                actorName = record["actor_name"]?.jsonPrimitive?.content ?: "",
                content = record["content"]?.jsonPrimitive?.content
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse notification", e)
            null
        }
    }

    private fun handleReply(payload: JsonObject) {
        val status = payload["status"]?.jsonPrimitive?.content
        Log.d(TAG, "Reply status: $status")
    }

    /**
     * Disconnect from realtime server.
     */
    fun disconnect() {
        heartbeatJob?.cancel()
        connectionJob?.cancel()
        session = null
        _connectionState.value = ConnectionState.DISCONNECTED
        Log.i(TAG, "Disconnected")
    }

    fun close() {
        disconnect()
        client.close()
    }
}

/**
 * Generic postgres change event.
 */
data class PostgresChange(
    val table: String,
    val type: String, // INSERT, UPDATE, DELETE
    val record: JsonObject?,
    val oldRecord: JsonObject?
)

/**
 * Parsed notification event for display.
 */
data class NotificationChange(
    val type: String, // comment, status_done, status_undone
    val caseNumber: String,
    val listSourceUrl: String,
    val actorName: String,
    val content: String?
)

File: ios/CourtLists/CourtLists/Services/RealtimeClient.swift
import Foundation

protocol RealtimeClientDelegate: AnyObject {
    func realtimeClient(_ client: RealtimeClient, didReceiveNotification notification: AppNotification)
    func realtimeClient(_ client: RealtimeClient, didReceiveCommentUpdate comment: Comment, eventType: String)
    func realtimeClient(_ client: RealtimeClient, didReceiveStatusUpdate status: CaseStatus, eventType: String)
    func realtimeClientDidConnect(_ client: RealtimeClient)
    func realtimeClientDidDisconnect(_ client: RealtimeClient)
}

class RealtimeClient: NSObject {
    weak var delegate: RealtimeClientDelegate?

    private var webSocket: URLSessionWebSocketTask?
    private var session: URLSession?
    private var userId: String?
    private var accessToken: String?
    private var isConnected = false
    private var heartbeatTimer: Timer?
    private var reconnectAttempts = 0
    private let maxReconnectAttempts = 5

    private let baseUrl = Configuration.realtimeUrl
    private let anonKey = Configuration.apiAnonKey

    override init() {
        super.init()
        session = URLSession(configuration: .default, delegate: self, delegateQueue: OperationQueue.main)
    }

    // MARK: - Connection Management

    func connect(userId: String, accessToken: String) {
        self.userId = userId
        self.accessToken = accessToken
        print("RealtimeClient: Connecting for user \(userId)")

        // Connect without token in URL (will send via message after connection)
        guard let url = URL(string: "\(baseUrl)?apikey=\(anonKey)&vsn=1.0.0") else {
            print("RealtimeClient: Invalid WebSocket URL")
            return
        }

        print("RealtimeClient: WebSocket connecting...")
        webSocket = session?.webSocketTask(with: url)
        webSocket?.resume()
        receiveMessage()
    }

    func disconnect() {
        heartbeatTimer?.invalidate()
        heartbeatTimer = nil
        webSocket?.cancel(with: .goingAway, reason: nil)
        webSocket = nil
        isConnected = false
    }

    // MARK: - Subscriptions

    private func subscribeToChannels() {
        guard let userId = userId else { return }

        print("RealtimeClient: Subscribing to channels for user \(userId)")

        // Subscribe to all tables in a single channel
        // Supabase Realtime v2 format
        let postgresChanges: [[String: Any]] = [
            [
                "event": "*",
                "schema": "public",
                "table": "notifications",
                "filter": "user_id=eq.\(userId)"
            ],
            [
                "event": "*",
                "schema": "public",
                "table": "comments"
            ],
            [
                "event": "*",
                "schema": "public",
                "table": "case_status"
            ]
        ]

        let payload: [String: Any] = [
            "event": "phx_join",
            "topic": "realtime:postgres_changes",
            "payload": [
                "config": [
                    "postgres_changes": postgresChanges
                ],
                "access_token": accessToken as Any
            ],
            "ref": UUID().uuidString
        ]

        sendMessage(payload)
        print("RealtimeClient: Subscription message sent")
    }

    // MARK: - Message Handling

    private func receiveMessage() {
        webSocket?.receive { [weak self] result in
            switch result {
            case .success(let message):
                switch message {
                case .string(let text):
                    self?.handleMessage(text)
                case .data(let data):
                    if let text = String(data: data, encoding: .utf8) {
                        self?.handleMessage(text)
                    }
                @unknown default:
                    break
                }
                self?.receiveMessage()

            case .failure(let error):
                print("WebSocket receive error: \(error)")
                self?.handleDisconnect()
            }
        }
    }

    private func handleMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            print("RealtimeClient: Failed to parse message: \(text.prefix(200))")
            return
        }

        let event = json["event"] as? String ?? ""
        let topic = json["topic"] as? String ?? ""
        print("RealtimeClient: Received event '\(event)' on topic '\(topic)'")

        switch event {
        case "phx_reply":
            if let payload = json["payload"] as? [String: Any],
               let status = payload["status"] as? String {
                print("RealtimeClient: phx_reply status: \(status)")
                if status == "ok" {
                    reconnectAttempts = 0
                    print("RealtimeClient: Channel subscribed successfully")
                } else if status == "error" {
                    if let response = payload["response"] as? [String: Any] {
                        print("RealtimeClient: Error response: \(response)")
                    }
                }
            }

        case "postgres_changes":
            handlePostgresChange(json, topic: topic)

        default:
            break // Don't log unhandled events (too noisy)
        }
    }

    private func handlePostgresChange(_ json: [String: Any], topic: String) {
        guard let payload = json["payload"] as? [String: Any],
              let data = payload["data"] as? [String: Any],
              let record = data["record"] as? [String: Any],
              let eventType = data["type"] as? String else {
            print("RealtimeClient: Failed to parse postgres change - payload structure issue")
            if let payload = json["payload"] {
                print("RealtimeClient: payload type: \(type(of: payload))")
            }
            return
        }

        // Get table name from the payload
        let table = data["table"] as? String ?? ""
        print("RealtimeClient: Change on table '\(table)', type: \(eventType)")
        print("RealtimeClient: Record keys: \(record.keys.sorted())")

        switch table {
        case "notifications":
            if let notification = parseNotification(record) {
                delegate?.realtimeClient(self, didReceiveNotification: notification)
            }
        case "comments":
            if let comment = parseComment(record) {
                delegate?.realtimeClient(self, didReceiveCommentUpdate: comment, eventType: eventType)
            }
        case "case_status":
            if let status = parseStatus(record) {
                delegate?.realtimeClient(self, didReceiveStatusUpdate: status, eventType: eventType)
            }
        default:
            print("RealtimeClient: Unknown table '\(table)'")
        }
    }

    // MARK: - Parsing

    private func parseNotification(_ record: [String: Any]) -> AppNotification? {
        guard let idNumber = record["id"] as? NSNumber,
              let userId = record["user_id"] as? String,
              let type = record["type"] as? String,
              let listSourceUrl = record["list_source_url"] as? String,
              let caseNumber = record["case_number"] as? String,
              let actorName = record["actor_name"] as? String,
              let createdAt = record["created_at"] as? String else {
            print("RealtimeClient: parseNotification failed - missing required fields")
            return nil
        }

        // Handle read as either Bool or Int
        var read = false
        if let readBool = record["read"] as? Bool {
            read = readBool
        } else if let readInt = record["read"] as? Int {
            read = readInt != 0
        }

        return AppNotification(
            id: idNumber.int64Value,
            userId: userId,
            type: type,
            listSourceUrl: listSourceUrl,
            caseNumber: caseNumber,
            caseTitle: record["case_title"] as? String,
            actorName: actorName,
            actorId: record["actor_id"] as? String,
            content: record["content"] as? String,
            read: read,
            createdAt: createdAt
        )
    }

    private func parseComment(_ record: [String: Any]) -> Comment? {
        guard let listSourceUrl = record["list_source_url"] as? String,
              let caseNumber = record["case_number"] as? String,
              let authorName = record["author_name"] as? String,
              let content = record["content"] as? String else {
            print("RealtimeClient: parseComment failed - missing required fields")
            print("RealtimeClient: record keys: \(record.keys)")
            return nil
        }

        // Handle id as NSNumber (JSON numbers aren't Int64 directly)
        var commentId: Int64? = nil
        if let idNumber = record["id"] as? NSNumber {
            commentId = idNumber.int64Value
        }

        // Handle urgent as either Bool or Int (database might send 0/1)
        var urgent = false
        if let urgentBool = record["urgent"] as? Bool {
            urgent = urgentBool
        } else if let urgentInt = record["urgent"] as? Int {
            urgent = urgentInt != 0
        }

        return Comment(
            id: commentId,
            listSourceUrl: listSourceUrl,
            caseNumber: caseNumber,
            userId: record["user_id"] as? String,
            authorName: authorName,
            content: content,
            urgent: urgent,
            createdAt: record["created_at"] as? String
        )
    }

    private func parseStatus(_ record: [String: Any]) -> CaseStatus? {
        guard let listSourceUrl = record["list_source_url"] as? String,
              let caseNumber = record["case_number"] as? String else {
            print("RealtimeClient: parseStatus failed - missing required fields")
            return nil
        }

        // Handle done as either Bool or Int
        var done = false
        if let doneBool = record["done"] as? Bool {
            done = doneBool
        } else if let doneInt = record["done"] as? Int {
            done = doneInt != 0
        }

        return CaseStatus(
            listSourceUrl: listSourceUrl,
            caseNumber: caseNumber,
            done: done,
            updatedBy: record["updated_by"] as? String,
            updatedAt: record["updated_at"] as? String
        )
    }

    // MARK: - Heartbeat

    private func startHeartbeat() {
        heartbeatTimer = Timer.scheduledTimer(withTimeInterval: 30, repeats: true) { [weak self] _ in
            self?.sendHeartbeat()
        }
    }

    private func sendHeartbeat() {
        let payload: [String: Any] = [
            "event": "heartbeat",
            "topic": "phoenix",
            "payload": [:],
            "ref": UUID().uuidString
        ]
        sendMessage(payload)
    }

    // MARK: - Sending

    private func sendMessage(_ payload: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let text = String(data: data, encoding: .utf8) else {
            return
        }

        webSocket?.send(.string(text)) { error in
            if let error = error {
                print("WebSocket send error: \(error)")
            }
        }
    }

    // MARK: - Reconnection

    private func handleDisconnect() {
        isConnected = false
        heartbeatTimer?.invalidate()
        heartbeatTimer = nil
        delegate?.realtimeClientDidDisconnect(self)

        if reconnectAttempts < maxReconnectAttempts {
            reconnectAttempts += 1
            let delay = Double(reconnectAttempts) * 2.0
            DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
                guard let self = self,
                      let userId = self.userId,
                      let accessToken = self.accessToken else { return }
                self.connect(userId: userId, accessToken: accessToken)
            }
        }
    }
}

// MARK: - URLSessionWebSocketDelegate

extension RealtimeClient: URLSessionWebSocketDelegate {
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol protocol: String?) {
        print("RealtimeClient: WebSocket opened")

        // Send access token to realtime:system
        if let token = accessToken {
            let tokenPayload: [String: Any] = [
                "event": "access_token",
                "topic": "realtime:system",
                "payload": ["access_token": token],
                "ref": UUID().uuidString
            ]
            sendMessage(tokenPayload)
            print("RealtimeClient: Sent access token")
        }

        // Directly subscribe to channels (skip phoenix join like Android)
        isConnected = true
        subscribeToChannels()
        startHeartbeat()
        delegate?.realtimeClientDidConnect(self)
    }

    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
        handleDisconnect()
    }
}
